broadinstitute/single_cell_portal_core
# Show "Differential expression" button whenever study has DE results (SCP-4954)

This shows the "Differential expression" button in more cases, helping discoverability now that we have 90% less DE data.

### Background <span id="background"></span>
Previously, we [showed the DE button](https://github.com/broadinstitute/single_cell_portal_core/pull/1502) only when the current clustering and annotation had DE results.  Then we [restricted DE](https://github.com/broadinstitute/single_cell_portal_core/pull/1715) to only cell type-like annotations, because the false positive rate of the Wilcoxon method seems too high even for exploratory analysis.  Naively, this would mean showing the DE button 90% less frequently.  That in turn would likely greatly diminish the overall value of DE in SCP.

Now, to A) only surface higher-quality DE while also B) avoiding a precipitous drop in DE discoverability and engagement, the big blue "Differential expression" button is shown whenever _any_ annotation in the study has DE results, even if the current annotation lacks DE results.  This means the DE button will be shown in _more_ cases than before the 90% restriction.

In that 90% case, the UX flow is:
1. See the DE button
2. Select a different annotation (and sometimes also a different clustering) from a list of those that have DE results
3. Select a group 
4. See the DE table

The main UX cost here is the the "Differential expression" button is more for a general feature of the study, rather than something directly related to the current clustering and annotation like the other controls in the right panel.  The bet here is that any cost in user confusion will be substantially less than the benefit of exploring DE results.

### Impact <span id="impact"></span>
The updated DE UX will likely increase indicators of interest (i.e. clicks on the DE button among users who see it), and may well also increase deeper DE engagement, like interacting with the DE table.

### Video <span id="video"></span>
Here's how it looks:

https://user-images.githubusercontent.com/1334561/222259046-5473589b-71cb-41bc-b384-4a5ba17864ac.mov

### Test <span id="test"></span>
A new automated test confirms behavior.  Optionally, if you want to manually verify, follow the steps in the video, i.e.:

1.  **If local**, after pulling, since we're only testing frontend changes, try booting with the frontend service worker cache enabled for faster page loads: `bin/docker-compose-setup.sh -c`.  Then go to a study with DE results.  **If on staging**, go to ["Human milk - differential expression"](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303/human-milk-differential-expression#study-visualize) (SCP303).
2. Change the default annotation to one that does not have "cell" and "type" in the name (e.g. `biosample_id`).  E.g. click "Settings" then "View options", then change "Default annotation" to e.g. "biosample_id", then click "Update settings", then go back to "Explore".
3. Confirm the blue "Differential expression" button appears, click it
4. Confirm the panel says "No DE results for:", and lists the current clustering and annotation, then gives a dropdown to select a clustering / annotation that has DE.  Select one of them.  If your current clustering lacks DE results, select one from the menu, then select an annotation.
6. Select a group
7. Confirm you see the DE table
8. Confirm the scatter plots show the name of your newly selected clustering and annotation

### Further details
[Slides](https://docs.google.com/presentation/d/1Zd_FO7HqUlxORJqTMEbp5SmZto-PvMKdtTg35nYecS4/edit) give more context.  Recordings for the [demo presentation](https://drive.google.com/file/d/1_6buLOdH4WLqUPMqFTC2QCQvH--DHCRp/view?t=51s) and [feedback](https://drive.google.com/file/d/1_6buLOdH4WLqUPMqFTC2QCQvH--DHCRp/view?t=15m02s) add richer commentary.

This satisfies SCP-4954.
